From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Sun, 21 Dec 2025 23:35:14 +0800
Subject: [PATCH 28/47] rtkit: Add version 10 support

Version 10 is used from iOS 10 to iOS 13. In version 10, EPMap is always
completed within 1 message. As a result, app endpoints has a different
offset, and each user of the rtkit API will need to be modified to properly
support version 10.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 src/rtkit.c | 42 ++++++++++++++++++++++++++++++++++--------
 src/rtkit.h |  2 ++
 2 files changed, 36 insertions(+), 8 deletions(-)

diff --git a/src/rtkit.c b/src/rtkit.c
index 14ad5b81..ff7c8a70 100644
--- a/src/rtkit.c
+++ b/src/rtkit.c
@@ -68,12 +68,15 @@
 #define MGMT_MSG_START_EP_IDX  GENMASK(39, 32)
 #define MGMT_MSG_START_EP_FLAG BIT(1)
 
-#define RTKIT_MIN_VERSION 11
+#define RTKIT_MIN_VERSION 10
 #define RTKIT_MAX_VERSION 12
 
 #define RTKIT_AKF_MSG_EP  GENMASK(63, 56)
 #define RTKIT_AKF_MSG_MSG GENMASK(55, 0)
 
+#define RTKIT_APP_EP_START_V10 0x5
+#define RTKIT_APP_EP_START_V11 0x20
+
 #define IOVA_MASK GENMASK(35, 0)
 
 enum rtkit_power_state {
@@ -103,6 +106,8 @@ struct rtkit_dev {
     iova_domain_t *dart_iovad;
     sart_dev_t *sart;
     bool sram;
+    u8 app_ep_start;
+    u8 protocol_ver;
 
     u64 dva_base;
 
@@ -318,6 +323,13 @@ rtkit_dev_t *rtkit_init(const char *name, const struct rtkit_iop_ops *iop_ops, v
     rtk->dart_iovad = dart_iovad;
     rtk->sart = sart;
     rtk->sram = sram;
+    /*
+     * treat only mgmt as system endpoint until more information
+     * can be obtained from mgmt's HELLO message
+     *
+     * (RTKIT_EP_MGMT = 0, so default app endpoint start is RTKIT_EP_CRASHLOG = 1)
+     */
+    rtk->app_ep_start = RTKIT_EP_CRASHLOG;
     rtk->iop_power = RTKIT_POWER_OFF;
     rtk->ap_power = RTKIT_POWER_OFF;
     rtk->dva_base = 0;
@@ -530,7 +542,7 @@ int rtkit_recv(rtkit_dev_t *rtk, struct rtkit_message *msg)
 
     while (rtk->iop_ops->recv(rtk, msg)) {
         /* if this is an app message we can just forward it to the caller */
-        if (msg->ep >= 0x20)
+        if (msg->ep >= rtk->app_ep_start)
             return 1;
 
         u32 msgtype = FIELD_GET(MGMT_TYPE, msg->msg);
@@ -682,6 +694,13 @@ bool rtkit_boot(rtkit_dev_t *rtk)
 
     rtkit_printf("booting with version %d\n", want_ver);
 
+    if (want_ver < 11)
+        rtk->app_ep_start = RTKIT_APP_EP_START_V10;
+    else
+        rtk->app_ep_start = RTKIT_APP_EP_START_V11;
+
+    rtk->protocol_ver = want_ver;
+
     msg.msg = FIELD_PREP(MGMT_TYPE, MGMT_MSG_HELLO_ACK);
     msg.msg |= FIELD_PREP(MGMT_MSG_HELLO_MINVER, want_ver);
     msg.msg |= FIELD_PREP(MGMT_MSG_HELLO_MAXVER, want_ver);
@@ -722,7 +741,7 @@ bool rtkit_boot(rtkit_dev_t *rtk)
             if (bitmap & (1U << i)) {
                 u8 ep_idx = 32 * base + i;
 
-                if (ep_idx >= 0x20)
+                if (ep_idx >= rtk->app_ep_start)
                     continue;
                 switch (ep_idx) {
                     case RTKIT_EP_CRASHLOG:
@@ -747,15 +766,17 @@ bool rtkit_boot(rtkit_dev_t *rtk)
             }
         }
 
-        if (msg.msg & MGMT_MSG_EPMAP_DONE)
+        if ((msg.msg & MGMT_MSG_EPMAP_DONE) || want_ver < 11)
             got_epmap = true;
 
         msg.msg = FIELD_PREP(MGMT_TYPE, MGMT_MSG_EPMAP_REPLY);
         msg.msg |= FIELD_PREP(MGMT_MSG_EPMAP_BASE, base);
-        if (got_epmap)
-            msg.msg |= MGMT_MSG_EPMAP_REPLY_DONE;
-        else
-            msg.msg |= MGMT_MSG_EPMAP_REPLY_MORE;
+        if (want_ver > 10) {
+            if (got_epmap)
+                msg.msg |= MGMT_MSG_EPMAP_REPLY_DONE;
+            else
+                msg.msg |= MGMT_MSG_EPMAP_REPLY_MORE;
+        }
 
         msg.ep = RTKIT_EP_MGMT;
 
@@ -866,3 +887,8 @@ bool rtkit_sleep(rtkit_dev_t *rtk)
     rtk->iop_ops->cpu_stop(rtk);
     return 0;
 }
+
+u8 rtkit_protocol_version(rtkit_dev_t *rtk)
+{
+    return rtk->protocol_ver;
+}
diff --git a/src/rtkit.h b/src/rtkit.h
index 4c57c924..c1d3737b 100644
--- a/src/rtkit.h
+++ b/src/rtkit.h
@@ -51,4 +51,6 @@ bool rtkit_unmap(rtkit_dev_t *rtk, u64 dva, size_t sz);
 bool rtkit_alloc_buffer(rtkit_dev_t *rtk, struct rtkit_buffer *bfr, size_t sz);
 bool rtkit_free_buffer(rtkit_dev_t *rtk, struct rtkit_buffer *bfr);
 
+u8 rtkit_protocol_version(rtkit_dev_t *rtk);
+
 #endif
